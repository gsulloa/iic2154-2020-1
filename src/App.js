import React, { useState, useRef, useEffect } from "react";
import styled, { css } from "styled-components";

const Header = styled.h1`
  color: red;
  ${props =>
    props.background &&
    css`
      background-color: ${props.background};
    `}
`;

const Row = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
`;

const Hello = props => <Header>Hello {props.name}</Header>;

export const Counter = ({ onSubmit = () => {} }) => {
  const [count, setCount] = useState(0);
  // state[0]: value
  // state[1]: setValue
  const increment = () => setCount(count + 1);
  const decrement = () => setCount(count - 1);
  const handleSubmit = () => {
    onSubmit(count);
    setCount(0);
  };
  return (
    <Row>
      <button className="" onClick={decrement}>
        -
      </button>
      <p data-testid="count">{count}</p>
      <button onClick={increment}>+</button>
      <button onClick={handleSubmit}>Enviar</button>
    </Row>
  );
};
const Countdown = props => {
  const [count, setCount] = useState(props.from);
  const interval = useRef();
  useEffect(() => {
    let i = props.from;
    interval.current = setInterval(() => {
      i = i - 1;
      setCount(i);
    }, 1000);
  }, [props.from]);
  return <p>{count}</p>;
};
function App() {
  const [numbers, setNumbers] = useState([]);
  const addNumbers = number => setNumbers([...numbers, number]);
  return (
    <div>
      <Hello name="Gabriel" />
      <Counter onSubmit={addNumbers} />
      <ul>
        {numbers.map((number, i) => {
          return (
            <li key={i}>
              <Countdown from={number} />
            </li>
          );
        })}
      </ul>
    </div>
  );
}

export default App;
