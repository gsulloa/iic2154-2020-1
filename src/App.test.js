import React from "react";
import { render, fireEvent } from "@testing-library/react";
import App, { Counter } from "./App";

test("renders learn react link", () => {
  const { getByText } = render(<App />);
  const linkElement = getByText(/hello/i);
  expect(linkElement).toBeInTheDocument();
});

describe("COunter", () => {
  it("Increment", () => {
    const fn = jest.fn();
    const { getByText, queryByText, getByTestId, baseElement } = render(
      <Counter onSubmit={fn} />
    );
    const decrementButton = getByText("-");
    const incrementButton = getByText("+");
    const submitButton = getByText(/enviar/i);
    expect(decrementButton).toBeInTheDocument();
    expect(incrementButton).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
    const undef = queryByText("no estoy");
    expect(undef).toBeNull();
    const count = getByTestId("count");
    expect(count).toBeInTheDocument();

    fireEvent.click(incrementButton);
    expect(count.innerHTML).toBe("1");
    fireEvent.click(incrementButton);

    expect(count.innerHTML).toBe("2");

    fireEvent.click(submitButton);
    expect(count.innerHTML).toBe("0");

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toBe(2);

    expect(baseElement).toMatchSnapshot();
  });
});
