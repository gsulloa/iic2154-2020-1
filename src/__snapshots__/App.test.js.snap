// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COunter Increment 1`] = `
<body>
  <div>
    <div
      class="sc-AxirZ jigHkS"
    >
      <button
        class=""
      >
        -
      </button>
      <p
        data-testid="count"
      >
        0
      </p>
      <button>
        +
      </button>
      <button>
        Enviar
      </button>
    </div>
  </div>
</body>
`;
